netkan/netkan/indexer.py

```diff
@@ -107,6 +101,10 @@ def status_attrs(self, new: bool = False) -> Dict[str, Any]:
             attrs['resources'] = resources
         if new:
             attrs['ModIdentifier'] = self.ModIdentifier
+        cache_path = self.ckan.cache_find_file
+        if cache_path:
+            cache_mtime = datetime.fromtimestamp(cache_path.stat().st_mtime)
+            attrs['last_downloaded'] = cache_mtime.astimezone(timezone.utc)
```

Comment:
My only note here is this will involve IO, require exposing the cache path to the indexer, and relying on file system timestamps. Which whilst not a big deal, the Inflator already knows when something is a cache hit or miss. Could we add this as a field on our SQS message? Is this significant effort on the inflator side of things?

Reply:
> involve IO

Well, for each mod, we'll have to:

1. Find the file
   Uses `Path.glob` to search the cache folder by filename. Unix filesystems list the file names contained in a directory in a dense sequence of contiguous lists (as many as will fit in a block), so the locality of reference here would be very high, one disk hit for a large number of file names (we might have around 3000-4000?). It doesn't load anything specific to the file until you ask for something beyond just the name. And these blocks are likely to be and stay cached.
   http://www.porcupine.org/forensics/forensic-discovery/chapter3.html#figure3.2
   ![image](https://user-images.githubusercontent.com/1559108/103811757-07362b00-5023-11eb-858f-3b540bb001cf.png)
2. Get the timestamp
   `stat`s the file and retrieves the mtime from the stat struct. I think this will be one disk access, for just the one file we found in the previous step.

So I don't expect a significant hit from this; it should be trivial next to all the other work the Infra does. I can look into gathering some measurements if you think that would be wise, though.

> require exposing the cache path to the indexer

Yup, I updated `prod-stack.py` to hopefully do this. Does that look like it will work?

> relying on file system timestamps

That's definitely what I want. ;)

> the Inflator already knows when something is a cache hit or miss

Hmm, I'd like to be able to backfill these values reasonably as well, which this PR's current code will handle for free by checking the currently cached files. The Inflator now has no idea when it last had to download each mod.

> Could we add this as a field on our SQS message? Is this significant effort on the inflator side of things?

There may be a technical quirk there. Netkan.exe has multiple places where it _might_ download a given file, and it checks the cache first each time (basically each translator or validator that wants to look at the ZIP asks the `CachingHttpService` to provide the file, independently, and it relies on the cache to avoid doing this repeatedly). So the first inflation of a new module would look like one cache miss followed by several cache hits. It feels like it would get messy to try to "remember" when those misses happen and when to consider them no longer applicable. (But I see reasons to avoid this approach anyway as discussed in the previous points.)

Reply:
I think that's a reasonable argument, we can always cross that bridge if we need to at a later date.